File: cpp/src/bench.cpp
// bench runs the word-search algorithm in-process (no HTTP), timing each
// canonical case per concurrency mode with warmup + median-of-N, and prints a
// single JSON report to stdout. Logs go to stderr.
//
// Modes:
//   file cases -> baseline (single thread), split (intra-file SPLIT_DEGREE chunks)
//   many cases -> baseline (sequential), fanout (per-length pool), nested (pool + split)
#include "search.h"

#include <nlohmann/json.hpp>

#include <algorithm>
#include <atomic>
#include <chrono>
#include <cstdlib>
#include <fstream>
#include <functional>
#include <iostream>
#include <string>
#include <thread>
#include <vector>

using json = nlohmann::json;

static int envInt(const char* key, int def) {
    const char* v = std::getenv(key);
    if (v && v[0]) {
        try {
            return std::stoi(v);
        } catch (...) {
        }
    }
    return def;
}

static std::vector<Hint> toHints(const json& arr) {
    std::vector<Hint> hints;
    for (const auto& h : arr) {
        Hint hint;
        hint.pos = h.value("pos", 0);
        if (h.contains("car") && !h["car"].is_null())
            hint.car = h["car"].get<std::string>();
        hint.inverted = h.value("inverted", false);
        hints.push_back(std::move(hint));
    }
    return hints;
}

struct Timing {
    double median_ms;
    double min_ms;
};

// timeMode warms up then times `iters` runs; returns count + timing.
static Timing timeMode(const std::function<std::vector<std::string>()>& fn,
                       int warmup, int iters, size_t& count) {
    for (int i = 0; i < warmup; i++) count = fn().size();
    std::vector<double> samples;
    samples.reserve(iters);
    for (int i = 0; i < iters; i++) {
        auto start = std::chrono::steady_clock::now();
        auto r = fn();
        auto elapsed = std::chrono::steady_clock::now() - start;
        samples.push_back(std::chrono::duration<double, std::milli>(elapsed).count());
        count = r.size();
    }
    std::sort(samples.begin(), samples.end());
    double median = samples[samples.size() / 2];
    if (samples.size() % 2 == 0)
        median = (samples[samples.size() / 2 - 1] + samples[samples.size() / 2]) / 2.0;
    return {median, samples.front()};
}

// runThroughput runs THROUGHPUT_OPS baseline scans with CONCURRENCY threads in
// flight, reporting aggregate ops/sec and median per-op latency under load.
static json runThroughput() {
    int concurrency = envInt("CONCURRENCY", 16);
    int ops = envInt("THROUGHPUT_OPS", 200);
    std::string lang = "fr";
    int nbCar = 11;
    std::vector<std::string> letters;
    for (char c = 'a'; c <= 'z'; c++) letters.push_back(std::string(1, c));
    Hint h;
    h.pos = 1;
    h.car = "x";
    h.inverted = false;
    std::vector<Hint> hints{h};

    inFile(lang, nbCar, letters, hints, false); // warmup

    std::vector<double> latencies(ops);
    std::atomic<int> idx{0};
    std::atomic<long> count{0};
    auto start = std::chrono::steady_clock::now();
    std::vector<std::thread> workers;
    for (int w = 0; w < concurrency; w++) {
        workers.emplace_back([&]() {
            for (;;) {
                int i = idx.fetch_add(1);
                if (i >= ops) return;
                auto t = std::chrono::steady_clock::now();
                auto r = inFile(lang, nbCar, letters, hints, false);
                latencies[i] = std::chrono::duration<double, std::milli>(
                                   std::chrono::steady_clock::now() - t).count();
                count.store(static_cast<long>(r.size()));
            }
        });
    }
    for (auto& t : workers) t.join();
    double elapsed = std::chrono::duration<double, std::milli>(
                         std::chrono::steady_clock::now() - start).count();
    std::sort(latencies.begin(), latencies.end());
    return json{
        {"workload", "file nb_car=11 pool=26 hint=1:x (baseline scan per op)"},
        {"concurrency", concurrency},
        {"ops", ops},
        {"elapsed_ms", elapsed},
        {"ops_per_sec", ops / (elapsed / 1000.0)},
        {"median_latency_ms", latencies[ops / 2]},
        {"count", count.load()},
    };
}

int main() {
    int warmup = envInt("BENCH_WARMUP", 20);
    int iters = envInt("BENCH_ITERS", 100);
    int degree = splitDegree();
    const char* cp = std::getenv("CASES_PATH");
    std::string casesPath = (cp && cp[0]) ? cp : "/app/cases.json";

    std::ifstream in(casesPath);
    if (!in.is_open()) {
        std::cerr << "cannot open cases: " << casesPath << std::endl;
        return 1;
    }
    json cases;
    in >> cases;

    json outCases = json::array();
    for (const auto& c : cases) {
        std::string name = c.value("name", "");
        std::string kind = c.value("kind", "file");
        std::string lang = c.value("lang", "fr");
        if (lang.empty()) lang = "fr";
        auto hints = toHints(c.value("lst_hint", json::array()));

        std::vector<std::pair<std::string, std::function<std::vector<std::string>()>>> modes;
        if (kind == "file") {
            int nbCar = c.value("nb_car", 0);
            auto letters = c.value("lst_car", std::vector<std::string>{});
            bool strict = c.value("strict", false);
            modes.push_back({"baseline", [=]() { return inFile(lang, nbCar, letters, hints, strict); }});
            modes.push_back({"split", [=]() { return inFileSplit(lang, nbCar, letters, hints, strict, degree); }});
        } else {
            std::string cars = c.value("cars", "");
            modes.push_back({"baseline", [=]() { return inManyFilesSeq(lang, cars, hints); }});
            modes.push_back({"fanout", [=]() { return inManyFiles(lang, cars, hints); }});
            modes.push_back({"nested", [=]() { return inManyFilesNested(lang, cars, hints, degree); }});
        }

        json modeJson = json::object();
        size_t count = 0;
        for (auto& [mname, fn] : modes) {
            size_t c2 = 0;
            Timing t = timeMode(fn, warmup, iters, c2);
            count = c2;
            modeJson[mname] = {{"median_ms", t.median_ms}, {"min_ms", t.min_ms}};
            std::cerr << "[C++] " << name << " / " << mname << ": " << c2
                      << " words, median " << t.median_ms << " ms" << std::endl;
        }
        outCases.push_back({{"name", name}, {"kind", kind}, {"count", count}, {"modes", modeJson}});
    }

    json throughput = runThroughput();
    std::cerr << "[C++] throughput: " << throughput["ops_per_sec"].get<double>()
              << " ops/s @ concurrency " << throughput["concurrency"].get<int>() << std::endl;

    json report = {
        {"language", "cpp"},
        {"label", "C++"},
        {"meta", {{"warmup", warmup}, {"iterations", iters}, {"split_degree", degree}}},
        {"cases", outCases},
        {"throughput", throughput},
    };
    std::cout << report.dump() << std::endl;
    return 0;
}

File: nest/src/bench.ts
// In-process benchmark for the Nest/Node implementation. Runs the algorithm
// directly (no HTTP), timing each canonical case per concurrency mode with
// warmup + median-of-N, and prints a single JSON report to stdout (logs go to
// stderr). Run as: node dist/bench.js
//
// Modes:
//   baseline -> synchronous scan on the main thread (no workers)
//   split    -> file split into SPLIT_DEGREE chunks across the worker pool
//   fanout   -> one task per word length across the pool (one whole-file scan)
//   nested   -> per-length AND per-chunk tasks queued on the fixed pool
import { readFileSync } from 'fs';
import { performance } from 'perf_hooks';
import { Hint, inFile, inManyFiles, planLengths } from './search/search';
import { WorkerPool } from './search/worker-pool';

interface RawHint {
  pos: number;
  car?: string | null;
  inverted?: boolean;
}
interface RawCase {
  name: string;
  kind: 'file' | 'many';
  lang?: string;
  nb_car?: number;
  lst_car?: string[];
  cars?: string;
  lst_hint?: RawHint[];
  strict?: boolean;
}

const WARMUP = parseInt(process.env.BENCH_WARMUP || '', 10) || 20;
const ITERS = parseInt(process.env.BENCH_ITERS || '', 10) || 100;
const DEGREE = Math.max(1, parseInt(process.env.SPLIT_DEGREE || '', 10) || 2);
const CASES_PATH = process.env.CASES_PATH || '/app/cases.json';

function toHints(raw: RawHint[] = []): Hint[] {
  return raw.map((h) => ({
    pos: h.pos,
    car: h.car ?? null,
    inverted: h.inverted ?? false,
  }));
}

async function timeMode(
  fn: () => Promise<string[]>,
): Promise<{ count: number; median_ms: number; min_ms: number }> {
  let count = 0;
  for (let i = 0; i < WARMUP; i++) count = (await fn()).length;
  const samples: number[] = [];
  for (let i = 0; i < ITERS; i++) {
    const start = performance.now();
    const r = await fn();
    samples.push(performance.now() - start);
    count = r.length;
  }
  samples.sort((a, b) => a - b);
  const mid = Math.floor(samples.length / 2);
  const median =
    samples.length % 2 === 0 ? (samples[mid - 1] + samples[mid]) / 2 : samples[mid];
  return { count, median_ms: median, min_ms: samples[0] };
}

async function main() {
  const pool = new WorkerPool();

  // Pool-based scan of one length, split into `chunkCount` chunks (merged in order).
  const runChunks = async (
    lang: string,
    length: number,
    letters: string[],
    hints: Hint[],
    strict: boolean,
    chunkCount: number,
  ): Promise<string[]> => {
    const chunks = await Promise.all(
      Array.from({ length: chunkCount }, (_, chunkIndex) =>
        pool.run({ lang, length, letters, hints, strict, chunkIndex, chunkCount }),
      ),
    );
    return chunks.flat();
  };

  const manyPool = async (
    lang: string,
    cars: string,
    hints: Hint[],
    chunkCount: number,
  ): Promise<string[]> => {
    const { minLen, maxLen, letters } = planLengths(cars, hints);
    if (maxLen < minLen) return [];
    const lengths: number[] = [];
    for (let l = maxLen; l >= minLen; l--) lengths.push(l);
    const perLength = await Promise.all(
      lengths.map((l) => runChunks(lang, l, letters, hints, false, chunkCount)),
    );
    return perLength.flat();
  };

  // Concurrent-load: THROUGHPUT_OPS whole-file scans dispatched to the worker
  // pool with CONCURRENCY in flight. Each op is one pool task (no intra-file
  // split) — the pool itself (default 2 workers) is the bottleneck under load.
  const runThroughput = async () => {
    const concurrency = parseInt(process.env.CONCURRENCY || '', 10) || 16;
    const ops = parseInt(process.env.THROUGHPUT_OPS || '', 10) || 200;
    const lang = 'fr';
    const nbCar = 11;
    const letters = 'abcdefghijklmnopqrstuvwxyz'.split('');
    const hints: Hint[] = [{ pos: 1, car: 'x', inverted: false }];
    await pool.run({ lang, length: nbCar, letters, hints, strict: false }); // warmup

    const latencies: number[] = new Array(ops);
    let count = 0;
    let next = 0;
    const start = performance.now();
    const worker = async () => {
      for (;;) {
        const i = next++;
        if (i >= ops) return;
        const t = performance.now();
        const r = await pool.run({ lang, length: nbCar, letters, hints, strict: false });
        latencies[i] = performance.now() - t;
        count = r.length;
      }
    };
    await Promise.all(Array.from({ length: concurrency }, () => worker()));
    const elapsed = performance.now() - start;
    latencies.sort((a, b) => a - b);
    return {
      workload: 'file nb_car=11 pool=26 hint=1:x (baseline scan per op)',
      concurrency,
      ops,
      elapsed_ms: elapsed,
      ops_per_sec: ops / (elapsed / 1000),
      median_latency_ms: latencies[Math.floor(ops / 2)],
      count,
    };
  };

  const cases: RawCase[] = JSON.parse(readFileSync(CASES_PATH, 'utf-8'));
  const outCases: unknown[] = [];

  for (const c of cases) {
    const lang = c.lang || 'fr';
    const hints = toHints(c.lst_hint);
    const modes: Record<string, () => Promise<string[]>> = {};

    if (c.kind === 'file') {
      const nbCar = c.nb_car ?? 0;
      const letters = c.lst_car ?? [];
      const strict = c.strict ?? false;
      modes.baseline = async () => inFile(lang, nbCar, letters, hints, strict);
      modes.split = () => runChunks(lang, nbCar, letters, hints, strict, DEGREE);
    } else {
      const cars = c.cars ?? '';
      modes.baseline = async () => inManyFiles(lang, cars, hints);
      modes.fanout = () => manyPool(lang, cars, hints, 1);
      modes.nested = () => manyPool(lang, cars, hints, DEGREE);
    }

    const modeJson: Record<string, { median_ms: number; min_ms: number }> = {};
    let count = 0;
    for (const [name, fn] of Object.entries(modes)) {
      const t = await timeMode(fn);
      count = t.count;
      modeJson[name] = { median_ms: t.median_ms, min_ms: t.min_ms };
      process.stderr.write(
        `[Nest] ${c.name} / ${name}: ${count} words, median ${t.median_ms.toFixed(4)} ms\n`,
      );
    }
    outCases.push({ name: c.name, kind: c.kind, count, modes: modeJson });
  }

  const throughput = await runThroughput();
  process.stderr.write(
    `[Nest] throughput: ${throughput.ops_per_sec.toFixed(1)} ops/s @ concurrency ${throughput.concurrency}\n`,
  );

  await pool.destroy();

  const report = {
    language: 'nest',
    label: 'Node/NestJS',
    meta: { warmup: WARMUP, iterations: ITERS, split_degree: DEGREE },
    cases: outCases,
    throughput,
  };
  process.stdout.write(JSON.stringify(report) + '\n');
}

main().catch((err) => {
  process.stderr.write(String(err && err.stack ? err.stack : err) + '\n');
  process.exit(1);
});
